elyra/util/operation_parser.py

```diff
@@ -0,0 +1,187 @@
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+
+import json
+import nbformat
+import re
+
+from abc import ABC, abstractmethod
+from jupyter_server.base.handlers import APIHandler
+from tornado import web
+from typing import Any, Type, TypeVar, List, Dict
+from ..util.http import HttpErrorMixin
+
+
+class OperationParserHandler(HttpErrorMixin, APIHandler):
+    """Handler to expose REST API to parse envs from a File"""
+
+    @web.authenticated
+    async def post(self):
+        msg_json = dict(title="Operation not supported.")
+        self.write(msg_json)
+        self.flush()
+
+    @web.authenticated
+    async def get(self, path=''):
+        path = path or ''
+
+        # TODO: Query and verify file type
+        # type = self.get_query_argument('type', default=None)
+        # if type not in {None, 'file', 'notebook'}:
+        #     raise web.HTTPError(400, u'Type %r is invalid' % type)
+
+        self.log.debug("Parsing file: %s", path)
+
+        model = await self.operation_parser(path)
+
+        self.set_status(200)
+        # TODO: Validation of model
+        self._finish_model(model)
+
+    async def operation_parser(self, operation_filepath):
+        """
+        Given the path to a File, will return a dictionary model
+        :param operation_filepath: absolute path to a File to be parsed
+        :return: a model dict
+        """
+        operation = OperationParser.get_instance(filepath=operation_filepath)
+        model = operation.model
+
+        return model
+
+    def _finish_model(self, model):
+        """Finish a JSON request with a model, setting relevant headers, etc."""
+        self.set_header('Last-Modified', model['last_modified'])
+        self.set_header('Content-Type', 'application/json')
+        self.finish(json.dumps(model))
+
+
+# Setup forward reference for type hint on return from class factory method.  See
+# https://stackoverflow.com/questions/39205527/can-you-annotate-return-type-when-value-is-instance-of-cls/39205612#39205612
+F = TypeVar('F', bound='OperationParser')
+
+
+class OperationParser(ABC):
+
+    @classmethod
+    def get_instance(cls: Type[F], **kwargs: Any) -> F:
+        """Creates an appropriate subclass instance based on the extension of the filepath"""
+        filepath = kwargs['filepath']
+        if '.ipynb' in filepath:
+            return NotebookOperationParser(filepath)
+        elif '.py' in filepath:
+            return PythonOperationParser(filepath)
+        elif '.r' in filepath:
+            pass
+            # return ROperatorParser(filepath)
+        else:
+            raise ValueError('Unsupported file type: {}'.format(filepath))
+
+    _type = None
+
+    def __init__(self, operation_filepath):
```

Comment:
The current implementation makes the parser `stateful and it really doesn't need to be. I was expecting something where I would say `foo.parse(path)` and this would simply do a `xxx.get_file_parser(path).parse()`